mantidproject/mantid
# Correct fractional binning in 2D workspaces (RebinnedOutput)

The handling of errors in workspaces with fractional bin weights (class `RebinnedOutput`) was found to be incorrect in #20599. This resolves that issue and also corrects code elsewhere which handled `RebinnedOutput` incorrectly (see below for details).

The major changes is in the code for `RebinnedOutput` and specifically the `rebinToFractionalOutput` method of `FractionalRebinning`. This affects three algorithms: `SofQWNormalisedPolygons` (where the bug was identified), `Rebin2D` and `ConvertToReflectometryQ` (using `ReflectometryTransform`). Users of these algorithms (principally powder inelastic neutron scattering and reflectometry) should check that they are still getting reasonable numbers out. (All the tests pass, but possibly not enough tests have been put in to cover all use cases).

In addition, the code for handling `RebinnedOutput` workspaces in the `Integration` algorithm has been corrected (it cast `RebinnedOutput` to `Workspace2D` and ignored the fractional bin weights), and that in `SumSpectra` has been verified to be correct. 

---

**Theory**

`rebinToFractionalOutput` works by taking an input quadrilateral (`inputQ`) and rebinning it into an output grid by sharing its count and errors amongst the new bins. First consider the case where `inputQ` is bigger than the output grid:

![text4172](https://user-images.githubusercontent.com/14106963/31711590-8fd7bb9a-b3f0-11e7-9a10-ad76d284e1a0.png)

For each output bin (regular grid squares), `rebinToFractionalOutput` calculates the overlap area `IntersectOverlap` of that bin with the input quadrilateral, and uses the fractional overlap area as a weight for the counts and errors. In the original (incorrect) code, it calculates for each bin:

```
signal_in_bin = signal_in_bin + ( signal_in_inputQ * weight )
variance_in_bin = variance_in_bin + ( error_in_inputQ * weight )**2
```

where `error` is the standard deviation and `variance=error**2`. Whilst the handling of the signal is correct, the handling of the errors is not, because the weights should be not be squared. To see this consider rebinning the data from the new grid back (regular squares) into the old grid. The error of the re-rebinned `inputQ` would then be:

```
new_variance_in_inputQ = sum_over_i ( weight_i**2 ) * error_in_inputQ**2
```

However, the sum over the squares of the *fractional* weights is **smaller** than unity - so successive rebinning from one grid to the other would yield progressively smaller errors. In fact the correct transformations should be:

```
signal_in_bin = signal_in_bin + ( signal_in_inputQ * weight )
variance_in_bin = variance_in_bin + ( error_in_inputQ**2 * weight )
```

e.g. it should be the variances and not the errors which should be weighted by the fractional area. Now the sum is over just the weights which will be unity.

Another problem with the implementation of `rebinToFractionalOutput` is that it does not "carry over the weights" when applied to a workspace which already has fractional bin weights (it treats all input workspaces as having unit weights). This means that if we were to rebin the previous workspace again into another grid with a different bin size which is does not fully cover the area of `inputQ`, e.g.:

![path4260](https://user-images.githubusercontent.com/14106963/31712547-518e8a28-b3f3-11e7-8cf0-d37df68af454.png)

then the `weight` which is used to scale the `signal` and `variance` will be wrong. In the above diagram, in the original code, the `weight` would be the ratio of the area of the whole of the blue square to the area of the green rectangle. The reweighted signal and error would then be an overestimate because the original data only covers the smaller area where the blue square and red quadrilateral intersects. Thus the correct weight should be:

```
new_weight = old_weight * intersectOverlap / inputQ
```

The final point is that in order to make the data appear sensible (accounting for the possibility that bin weights may vary by a lot in nearby bins which would otherwise make the signal/error appear to vary by alot between adjacent bins), the data in the workspace is `finalized` by dividing by the weights at the end of most algorithms. In this case, because the data is just scaled, the code correctly divides the signal and *error* (not variance) by the weights (it's not added in quadrature). However, for subsequent processing (e.g. in `Rebin2D` after `SofQWNormalisedPolygon`), the data should be `unfinalized` or treated in a different way which accounts for the fact that it has been scaled by the weights. This was not done in `Rebin2D` or `Integration`.

---

**To test:**

Use the code in issue #20599 and check that the rebinned outputs for both coarse and fine Q bin steps now agree exactly in signal and errors. 

Run a standard inelastic neutron powder reduction (e.g. from #15953) and check that everything still works as usual and the data appears ok.

Run a standard reflectometry reduction (e.g. from #15188 and its PR) and check that everything still looks ok.

Fixes #20599. 

**Release Notes** 

To follow.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
